proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+use of level information to accept more programs, but this seems like an
+undesirable level of complexity.  Thus the following example is rejected::
+
+  module X where
+    data X = MkX
+
+  module Normal where
+    import X
+    instance Show X where show _ = "normal"
+
+  module Splice where
+    import X
+    instance Show X where show _ = "splice"
+
+  module Bottom where
+    import X (X(..))
+    import splice X (X(..))
+    import Normal ()        -- imports instance Show X at level 0
+    import splice Splice () -- imports a different instance Show X at level -1
+
+    s1 = show MkX -- Error: overlapping instances defined in ``Normal`` and ``Splice``
+
+However the following is accepted::
+
+  module X where
+    data X = MkX deriving Show
+
+  module Bottom where
+    import X (X(..))        -- imports instance Show X at level 0
+    import splice X (X(..)) -- imports the same instance Show X at level -1
+    import splice Language.Haskell.TH.Lib ( stringE )
+
+    s1 = show MkX                 -- Uses instance at level 0
+    s2 = $( stringE (show MkX) )  -- Uses instance at level -1
+
+
+Exports of class instances
+##########################
+
+Only instances available at level 0 are re-exported from a module.  For example,
+the following is rejected::
+
+  module X where
+    data X = MkX
+
+  module Splice where
+    import X
+    instance Show X where show _ = "splice"
+
+  module Y where
+    import splice Splice () -- imports instance Show X at level -1
+
+  module Bottom where
+    import X (X(..))
+    import Y ()
+
+    s1 = show MkX -- Error: no instance for Show X
+
+Even though ``Y`` has access to the instance at level -1, it does not re-export it.  Thus ``Bottom`` does not import the instance.
+
+This is necessary for a clean separation between stages, because instances may exist only at compile-time or only at runtime, just like identifiers.
+
+
+
+Examples
+--------
+
+Splice imports
+##############
+
+A "splice" import is prefixed with ``splice``. In this example, identifiers from
+``A`` can be used only in top-level splices and identifiers from ``B``
+cannot be used in quotes or splices::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+  {-# LANGUAGE TemplateHaskell #-}
+  module Main where
+
+  import splice A (foo)  -- foo :: Int -> Q Exp
+  import B (bar)         -- bar :: Int -> Q Exp
+
+  x = $(foo 25) -- Accepted
+  y = $(bar 33) -- Error: bar imported at level 0 but used at level -1
+
+Thus:
+
+1. When compiling module ``Main``, even though ``TemplateHaskell`` is enabled,
+   only identifiers from module ``A`` will be used in top-level splices so
+   only ``A`` (and its dependencies) needs to compiled to object code before starting to compile ``Main``.
+2. When cross-compiling, ``A`` needs to be built only for the host and ``B``
+   only for the target.
+
+If the same module is needed to be used at different levels
+then two import declarations can be used::
```

Comment:
Please put this very important point in the Change Specification, not in Examples.